test/integration/examples/apiserver_test.go

```diff
@@ -17,71 +17,316 @@ limitations under the License.
 package apiserver
 
 import (
-	"crypto/tls"
+	"crypto/x509"
 	"encoding/json"
 	"fmt"
 	"io/ioutil"
-	"net/http"
+	"net"
 	"os"
+	"path"
+	"strconv"
+	"strings"
 	"testing"
 	"time"
 
-	"github.com/golang/glog"
 	"github.com/stretchr/testify/assert"
 
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	client "k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/clientcmd"
 	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	"k8s.io/client-go/util/cert"
+	apiregistrationv1alpha1 "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1alpha1"
+	aggregatorclient "k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset"
+	kubeaggregatorserver "k8s.io/kube-aggregator/pkg/cmd/server"
+	"k8s.io/kubernetes/cmd/kube-apiserver/app"
+	"k8s.io/kubernetes/cmd/kube-apiserver/app/options"
 	"k8s.io/kubernetes/test/integration/framework"
 	"k8s.io/sample-apiserver/pkg/apis/wardle/v1alpha1"
-	"k8s.io/sample-apiserver/pkg/cmd/server"
+	sampleserver "k8s.io/sample-apiserver/pkg/cmd/server"
 )
 
-const securePort = "6444"
-
 var groupVersion = v1alpha1.SchemeGroupVersion
 
 var groupVersionForDiscovery = metav1.GroupVersionForDiscovery{
 	GroupVersion: groupVersion.String(),
 	Version:      groupVersion.Version,
 }
 
-func TestRunServer(t *testing.T) {
-	masterConfig := framework.NewIntegrationTestMasterConfig()
-	_, s := framework.RunAMaster(masterConfig)
-	defer s.Close()
+// It seems like this should work, but it doesn't
+func localPort() (int, error) {
+	l, err := net.Listen("tcp", ":0")
+	if err != nil {
+		return 0, err
+	}
+	defer l.Close()
+	addr := strings.Split(l.Addr().String(), ":")
+	port, err := strconv.Atoi(addr[len(addr)-1])
+	if err != nil {
+		return 0, err
+	}
+	return port, nil
+}
+
+func TestAggregatedAPIServer(t *testing.T) {
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+
+	kubePort, err := localPort()
```

Comment:
@sttts Are you thinking that it should be possible to use a zero port in starting a server?  I originally thought that might be possible, since ``runServer`` appears to return the effective port, but got tripped up by the informer initialization.  

Taking a look at ``runServer`` again, it looks like the returned port is not guaranteed to be the same as the port that is bound to if the zero port is specified.  Why is the effective port determined in runServer at all?  The way it's written it seems like parsing the address would be simpler.